crates/cow-amm/src/indexer.rs

```diff
@@ -0,0 +1,162 @@
+use {
+    crate::{
+        cow_amm::CowAmm,
+        cow_amm_constant_product_factory::CowAmmConstantProductFactoryHandler,
+    },
+    contracts::cow_amm_constant_product_factory,
+    ethcontract::{common::DeploymentInformation, dyns::DynWeb3, Address},
+    ethrpc::current_block::RangeInclusive,
+    shared::event_handling::EventStoring,
+    std::{
+        collections::{BTreeMap, HashMap},
+        sync::Arc,
+    },
+    tokio::sync::RwLock,
+};
+
+pub type CowAmmRegistry = HashMap<Address, CowAmm>;
+/// Registry with the format: (block number, (log index, event))
+pub type EventsRegistry = BTreeMap<u64, BTreeMap<usize, cow_amm_constant_product_factory::Event>>;
+
+/// CoW AMM indexer which stores events in-memory.
+#[derive(Clone)]
+pub struct Indexer {
+    cow_amms: Arc<RwLock<CowAmmRegistry>>,
+    events_registry: Arc<RwLock<EventsRegistry>>,
+    pub first_block: u64,
+}
+
+impl Indexer {
+    pub async fn new(
+        web3: &DynWeb3,
+        cow_amm_factory_contract: Option<&cow_amm_constant_product_factory::Contract>,
+    ) -> Self {
+        let cow_amm_constant_product_factory =
+            if let Some(cow_amm_factory_contract) = cow_amm_factory_contract {
+                cow_amm_factory_contract.clone()
+            } else {
+                contracts::CowAmmConstantProductFactory::deployed(web3)
+                    .await
+                    .expect("Failed to find deployed CowAmmConstantProductFactory")
+            };
+        let first_block = Self::first_block(cow_amm_constant_product_factory)
+            .await
+            .unwrap_or(0);
```

Comment:
I would move the `.unwrap_or(0)` into `Self::first_block()`. IMO it's also fine to panic in initialization code that only runs on start up anyway to make the error handling and APIs simpler.

Reply:
@MartinquaXD moved it! thanks!

The problem with the panic here is that for e2e tests when we instantiate a new factory contract, it doesn't contain deployment information, so that's why I set it to zero for that case.